jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * (C) Copyright 2017-2017, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.tour;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.interfaces.TSPAlgorithm;
+import org.jgrapht.graph.GraphWalk;
+
+import java.util.*;
+
+/**
+ * A dynamic programming algorithm for the TSP problem.
+ *
+ * <p>
+ * The travelling salesman problem (TSP) asks the following question: "Given a list of cities and
+ * the distances between each pair of cities, what is the shortest possible route that visits each
+ * city exactly once and returns to the origin city?".
+ *
+ * <p>
+ * This is an implementation of the Held-Karp algorithm which returns a minimum-cost Hamiltonian tour.
+ * The implementation requires the input graph to contain at least one vertex.
+ * The running time is $O(2^{|V|} \times |V|^2)$ and it takes $O(2^{|V|} \times |V|)$ extra memory.
+ *
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">wikipedia</a> for more
+ * details about TSP.
+ *
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Held%E2%80%93Karp_algorithm">wikipedia</a> for more
+ * details about the dp algorithm.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class HeldKarpTSP<V, E>
+        implements TSPAlgorithm<V, E> {
+
+    /**
+     * Construct a new instance
+     */
+    public HeldKarpTSP() {
+    }
+
+    private void printState(int state, int n){
```

Comment:
Consider deleting this, or moving this to the test package if really needed for debugging.